Fix: `gh gist view` and `gh gist edit` prompts with no TTY

mateusmarquezini opened the pull request:
Fixes #10042

mateusmarquezini committed:
fixing gh gist view prompts with no TTY

mateusmarquezini committed:
Merge branch 'trunk' into fix/issue_10042

mateusmarquezini committed:
#10042 fixed test

mateusmarquezini committed:
Merge branch 'fix/issue_10042' of github.com:mateusmarquezini/cli into fix/issue_10042

mateusmarquezini committed:
Merge branch 'trunk' into fix/issue_10042

mateusmarquezini committed:
#10042: Add error messages for 'gh gist view/edit' prompts when no TTY is detected

mateusmarquezini committed:
#10042 removed unnecessary field

BagToad commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -108,6 +108,9 @@ func editRun(opts *EditOptions) error {
 	if gistID == "" {
 		cs := opts.IO.ColorScheme()
 		if gistID == "" {
+			if !opts.IO.CanPrompt() {
+				return cmdutil.FlagErrorf("gist ID or URL required when not running interactively")
+			}
```
Nit: I think we should keep the spacing consistent between the changes to `gist edit` and `gist view` 😄  

```suggestion
			if !opts.IO.CanPrompt() {
				return cmdutil.FlagErrorf("gist ID or URL required when not running interactively")
			}
			
```

BagToad commented on pkg/cmd/gist/view/view_test.go:
```diff
@@ -157,7 +158,7 @@ func Test_viewRun(t *testing.T) {
 					},
 				},
 			},
-			wantOut: "test interactive mode\n",
+			wantErr: true,
```
Thank you for fixing these tests @mateusmarquezini ✨ 

I think this test should be fixed in a different way 🤔  This test was originally testing that interactivity works, and I think we should preserve that meaning instead of checking that it errors now. I could see value in an _additional_ test that checks that we get the new error message when no argument is provided and no TTY is available as expected.

Here's an example test from `gh release create` 

https://github.com/cli/cli/blob/4c2ba5681d4777dc364bad989eda6a68189ca6a5/pkg/cmd/release/create/create_test.go#L68-L74

To do this, we would need to copy this logic by adding an `isTTY` value to the table test struct definition ([example](https://github.com/cli/cli/blob/4c2ba5681d4777dc364bad989eda6a68189ca6a5/pkg/cmd/release/create/create_test.go#L42-L49)):

https://github.com/cli/cli/blob/4c2ba5681d4777dc364bad989eda6a68189ca6a5/pkg/cmd/release/create/create_test.go#L42-L49

Then initialize I/O based on that in the test logic ([example](https://github.com/cli/cli/blob/4c2ba5681d4777dc364bad989eda6a68189ca6a5/pkg/cmd/release/create/create_test.go#L353-L366)):

https://github.com/cli/cli/blob/4c2ba5681d4777dc364bad989eda6a68189ca6a5/pkg/cmd/release/create/create_test.go#L353-L366

I think we should add this test to the `gh gist edit` changes as well 🙏 

----

To summarize, I'd like both `gh gist edit` and `gh gist view` tests to contain at least one test that checks that interactivity works when a TTY is available. I think it would also be good if both commands had one test to check that we receive an error message when TTY is _not_ available.

`gh release view` seems to be a good command to look into for code inspiration as the same patterns exist there. A code search for `required when not running interactively` in all `_test.go` files also appears to be helpful for finding additional inspiration.

I recognize this increases the complexity of this PR a bit; please let me know if you have any questions @mateusmarquezini ❤️

BagToad reviewed (changes requested):
Thank you for your patience with this one @mateusmarquezini ✨ 

I wasn't sure if you're ready for review on this yet per your PR description, but I've gone ahead and added some comments in case that is useful.

I think the implementation is sound, but I think we should improve the tests to align more with how some other commands are testing for this interactivity error message - I've put more details about that in the PR comments.

Please let me know if you have any questions or when you're ready for another review ❤️

mateusmarquezini commented:
Hey @BagToad! Thanks so much for your amazing review & suggestions! 🌟

Oops, I forgot to update the PR description—thanks for the heads up!

About your other comments—awesome! I'll apply those suggestions and ping you once it's ready for another review!

Have an awesome weekend! ✨

mateusmarquezini committed:
Merge branch 'trunk' into fix/issue_10042

mateusmarquezini committed:
Update pkg/cmd/gist/edit/edit.go

Update pkg/cmd/gist/edit/edit.go after code review

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

mateusmarquezini committed:
removed unnecessary space

mateusmarquezini committed:
#10042: Applied code review suggestions to view_test.go

mateusmarquezini committed:
#10042: Attempt to add a new test scenario for handling error messages when TTY is unavailable

mateusmarquezini committed:
Merge branch 'trunk' into fix/issue_10042

mateusmarquezini commented:
Hey @BagToad! Thanks again for all your suggestions—they were super helpful and really guided me through the corrections!✨ 

I've got some questions about the fixes in `gist/edit/edit_test.go`. Maybe you can help me out!

So, I tried applying the same approach I used in `view_test.go` ([example](https://github.com/mateusmarquezini/cli/blob/fix/issue_10042/pkg/cmd/gist/view/view_test.go#L168-L172)) to `edit_test.go` ([example](https://github.com/mateusmarquezini/cli/blob/4eb0e8138f5650061127e89abdc7f9d640483f6c/pkg/cmd/gist/edit/edit_test.go#L522-L529)), as it'd be great to test the scenario where TTY isn't available.

However, I'm running into [this](https://github.com/cli/cli/actions/runs/12749571145/job/35532310383?pr=10048#step:5:92) error and feeling a bit stuck on the right way to fix this specific test.

Looking at the code, it seems the issue happens because the test runner always sets `tt.opts.Selector = "1234'` ([example](https://github.com/mateusmarquezini/cli/blob/4eb0e8138f5650061127e89abdc7f9d640483f6c/pkg/cmd/gist/edit/edit_test.go#L566)), which makes it reach the API call instead of failing early. Fixing this would require some additional changes I wasn't expecting 😂, especially in the HTTP stubs for `edit_test.go`.

My question is: Should I keep going with this approach, or do you think there's an easier way to handle this?

Additionally, feel free to give me some feedback on the [test](https://github.com/cli/cli/pull/10048/commits/977e2326a25d48c7fb4250b2927760ca28f18a79#diff-511b0374c8e5881fa4fd71ab5198a6ecb9c0f87362697b3c58c4bb2d5426e57f) I already created in `pkg/cmd/gist/view/view_test.go`, since, as I mentioned, I'm applying a similar test in `gist/edit/edit_test.go` and I'd love your feedback before moving forward.

Thanks in advance for your input! 🌟

BagToad committed:
Fix: Gist edit tests for interactivity

This changes the gist edit tests to use the positive `istty` instead of the previous inverse `nontty`, which is consistent with the way other commands are written.

BagToad commented:
Hey @mateusmarquezini! 👋 

> However, I'm running into [this](https://github.com/cli/cli/actions/runs/12749571145/job/35532310383?pr=10048#step:5:92) error and feeling a bit stuck on the right way to fix this specific test.
> 
> Looking at the code, it seems the issue happens because the test runner always sets tt.opts.Selector = "1234' ([example](https://github.com/mateusmarquezini/cli/blob/4eb0e8138f5650061127e89abdc7f9d640483f6c/pkg/cmd/gist/edit/edit_test.go#L566)), which makes it reach the API call instead of failing early. Fixing this would require some additional changes I wasn't expecting 😂, especially in the HTTP stubs for edit_test.go.
> 
> My question is: Should I keep going with this approach, or do you think there's an easier way to handle this?

I see what you mean! These `gist edit` tests don't appear to be consistent with other commands in the CLI, especially `gist view`, but they really should be. 

So, to save us the headache trying to work around that inconsistency, I'm going to push some commits to this PR to fix up the tests, make them easier to read, and hopefully more consistent - I hope that is okay with you 😃 

I'm also resolving some merge conflicts that popped up 😜

BagToad committed:
Merge 'trunk' into fix/issue_10042

BagToad committed:
Fix: gist edit/view tests var name consistency

BagToad committed:
Fix: gist edit test name

BagToad commented:
> Looking at the code, it seems the issue happens because the test runner always sets tt.opts.Selector = "1234' ([example](https://github.com/mateusmarquezini/cli/blob/4eb0e8138f5650061127e89abdc7f9d640483f6c/pkg/cmd/gist/edit/edit_test.go#L566)), which makes it reach the API call instead of failing early. Fixing this would require some additional changes I wasn't expecting 😂, especially in the HTTP stubs for edit_test.go.

To be more specific with what I changed in regard to this - I removed that hardcoded selector assignment that was causing us problems and instead added it to the tests that require it. It's a bit more repetitive, but I do not currently see a better way of doing it because we have a situation where we _want_ the selector to be nothing 😁 

I added comments to some of the HTTP stubs and parameter checking that should hopefully make this easier to follow. I also renamed some of the table test variables to make it clearer what they're meant for. 

I also added some _extra_ PR comments to help you, me, and future readers understand what's going on in these tests. Feel free to click "resolve" on those comments at your discretion 👍

BagToad committed:
Fix: accidental whitespace in gist edit

BagToad commented on pkg/cmd/gist/edit/edit_test.go:
```diff
@@ -519,28 +549,74 @@ func Test_editRun(t *testing.T) {
+		}
+
+		if tt.mockGistList {
+			sixHours, _ := time.ParseDuration("6h")
+			sixHoursAgo := time.Now().Add(-sixHours)
+			reg.Register(httpmock.GraphQL(`query GistList\b`),
+				httpmock.StringResponse(
+					fmt.Sprintf(`{ "data": { "viewer": { "gists": { "nodes": [
+							{
+								"description": "whatever",
+								"files": [{ "name": "cicada.txt" }, { "name": "unix.md"	}],
+								"isPublic": true,
+								"name": "1234",
+								"updatedAt": "%s"
+							}
+							],
+							"pageInfo": {
+								"hasNextPage": false,
+								"endCursor": "somevaluedoesnotmatter"
+							} }	} }	}`, sixHoursAgo.Format(time.RFC3339))))
 			reg.Register(httpmock.REST("GET", "gists/1234"),
-				httpmock.JSONResponse(tt.gist))
+				httpmock.JSONResponse(tt.mockGist))
 			reg.Register(httpmock.GraphQL(`query UserCurrent\b`),
 				httpmock.StringResponse(`{"data":{"viewer":{"login":"octocat"}}}`))
+
+			gistList := "cicada.txt whatever about 6 hours ago"
+			pm.RegisterSelect("Select a gist",
+				[]string{gistList},
+				func(_, _ string, opts []string) (int, error) {
+					return prompter.IndexFor(opts, gistList)
+				})
```
This was needed to properly test interactivity in the case of "no gist selector provided" because when that happens a list of gists is fetched and we need the HTTP and selector mocks for that request.

BagToad replied:
Aligned as closely to `gist view` tests as possible 😃

BagToad commented on pkg/cmd/gist/edit/edit_test.go:
```diff
@@ -519,28 +549,74 @@ func Test_editRun(t *testing.T) {
+			tt.opts = &EditOptions{}
+		}
+
+		if tt.opts.Selector != "" {
+			// Only register the HTTP stubs for a direct gist lookup if a selector is provided.
+			if tt.mockGist == nil {
+				// If no gist is provided, we expect a 404.
+				reg.Register(httpmock.REST("GET", fmt.Sprintf("gists/%s", tt.opts.Selector)),
+					httpmock.StatusStringResponse(404, "Not Found"))
+			} else {
+				// If a gist is provided, we expect the gist to be fetched.
+				reg.Register(httpmock.REST("GET", fmt.Sprintf("gists/%s", tt.opts.Selector)),
+					httpmock.JSONResponse(tt.mockGist))
+				reg.Register(httpmock.GraphQL(`query UserCurrent\b`),
+					httpmock.StringResponse(`{"data":{"viewer":{"login":"octocat"}}}`))
+			}
+		}
```
When we have a selector provided, we expect the command to fetch that gist. If the test has defined a mock gist, then we return that. Otherwise we consider that a 404 and return that instead.

BagToad commented:
> Additionally, feel free to give me some feedback on the [test](https://github.com/cli/cli/pull/10048/commits/977e2326a25d48c7fb4250b2927760ca28f18a79#diff-511b0374c8e5881fa4fd71ab5198a6ecb9c0f87362697b3c58c4bb2d5426e57f) I already created in pkg/cmd/gist/view/view_test.go, since, as I mentioned, I'm applying a similar test in gist/edit/edit_test.go and I'd love your feedback before moving forward.

Looks good to me! I'm sorry that `gist edit` wasn't as straightforward 😆 

@mateusmarquezini I'll let you have a look over the changes I've pushed - if you have any questions, let me know. Otherwise, let me know and I'll do a last review 🎉

mateusmarquezini commented:
Thank you so much for the amazing help you provided @BagToad!❤️

I honestly didn't expect all those changes in the `gist edit` haha—thank you so much for your contribution to this PR!

Feel free to go ahead and make a final review on it 🎉

BagToad reviewed (approved):
LGTM 🚀 

```
❯ ./bin/gh gist view | cat
gist ID or URL required when not running interactively

Usage:  gh gist view [<id> | <url>] [flags]

Flags:
  -f, --filename string   Display a single file from the gist
      --files             List file names from the gist
  -r, --raw               Print raw instead of rendered gist contents
  -w, --web               Open gist in the browser
  
❯ ./bin/gh gist edit | cat
gist ID or URL required when not running interactively

Usage:  gh gist edit {<id> | <url>} [<filename>] [flags]

Flags:
  -a, --add string        Add a new file to the gist
  -d, --desc string       New description for the gist
  -f, --filename string   Select a file to edit
  -r, --remove string     Remove a file from the gist
  
```
